### PR TITLE
Expand finite volume support

### LINKED ISSUE #15063: FV: add support for cross-diffusion terms for non-othogonal meshes
## Reason
Non-orthogonal meshes cause distortion in FV diffusion terms unless you calculate special correction factors.  Implementing this will improve accuracy of FV results for unstructured (non-orthogonal) meshes.

## Design
The correction factors only change when the mesh changes - so they can be computed+cached up front and only updated when the mesh moves.  These correction factors should probably be stored and accessed via the face info objects for their respective faces.

## Impact
This should not affect existing APIs - it will just increase the range of scenarios where FV gives good solution accuracy/correctness.

### LINKED ISSUE #15066: FV: add support for higher-order solution reconstruction
## Reason
Various numerical flux methods and other FV bells-and-whistles benefit from or require having gradient approximations for mesh cells (i.e. `_grad_u_left` and `_grad_u_right` in fv kernels).  

## Design
Computing these gradients requires values from a cell's neighbors.  This calculation should probably be initiated on-demand for FV variables when they are computing/setting the rest of the solution, gradient, etc. values.  Then the variable could provide these values to kernels, bcs, etc. transparently just like objects usually get gradients.  The FV variable could cash this value for the next time it is needed (since each cell's gradient info will be needed for every visit to a face of that cell).  We will need to create a new "reconstruction" object that users can write custom versions of and enable/swap from input file syntax.

## Impact
Should not affect existing APIs

### LINKED ISSUE #15329: ADMaterial coupled to AuxVariables with FV gives segmentation fault
## Bug Description
Using FV, having a diffusion coefficient depending on a coupled auxiliary variable gives a segmentation fault. If we couple the material to a FV variable, segmentation fault occurs only if the coupled variable is not governed by a FVDiffusion kernel.

## Steps to Reproduce
Input files describing the issue will be submitted

## Impact
Material properties depending on auxiliary variables is not possible. Fixing this would increase a lot the flexibility of the FV support.

### LINKED ISSUE #15836: FV: Support use of finite volume entities with the aux system
## Reason
A user should be able to use all finite volume variables in their input file and expect it to work seamlessly with the auxiliary system

## Design
Generalize API calls in `AuxKernel` and its derivatives

## Impact
Improve finite volume user experience with the auxiliary system

### LINKED ISSUE #15894: Block Restriction with FV crashes sometimes
## Reason
This feature is needed to limit some of the physics (e.g. solid heat conduction) only where is needed.

## Design
The same design we are using for FE would work.

## Impact
It will improve the usability of the FV kernels and materials.

Currently when Block Restriction is applied the problem error out with a segfault (see attached file for testing).
[block_restriction.txt](https://github.com/idaholab/moose/files/5321990/block_restriction.txt)

### PR DESCRIPTION
**Pull request by lindsayad**
Includes:
- Gradient reconstruction
- Non-orthogonal corrections
- Auxiliary finite volume variable support
- Verification of FV for advection-diffusion-reaction type problems

Remaining to-do:

- [x] Verify that "non-local" computation of gradient works with block restriction

Closes #15063 #15836 #15066 #15825 #15329

Refs #15062, e.g. this adds support for coupling FV into FE

### CODE REVIEW

**Comment by lindsayad** [comment]
@rwcarlsen With 115 commits in the INSFV branch it was a little time consuming extricating only the framework commits that would support gradient reconstruction testing. So here you go in one big squashed commit all the framework level changes from INSFV support. I was actually pleased to see that while this is 5k lines, there's actually not nearly that much in terms of framework changes. Most of the lines are from testing

**Review by rwcarlsen — CHANGES_REQUESTED** [review]
Here is my first pass through.  Mostly questions and minor nitpicks.  Could you provide maybe some suggestions of themes or bigger things/changes for me to focus down on for the review?  That would help make a second pass more effective.

File: framework/include/mesh/MooseMesh.h
Comment (lindsayad) [reply framework/include/mesh/MooseMesh.h:988]: Because of the addition of `_all_face_info`

**Comment by lindsayad** [comment]
> Could you provide maybe some suggestions of themes or bigger things/changes for me to focus down on for the review?

I think the most meaningful code to look over are the methods added at the end of `MooseVariableFV`. Another design question is whether `FVFaceInterface` should exist at all. My vote would be to move everything into `FVUtils` which is what I was gradually doing over the course of the INSFV work. I was thinking about whether to create an `FV` namespace nested inside of the `Moose` namespace. What are your thoughts on that? I could go either way.

File: framework/include/mesh/MooseMesh.h
Comment (lindsayad) [reply framework/include/mesh/MooseMesh.h:988]: I think I'm going to actually convert this to a container of `const FaceInfo *` and encapsulate any modification of `FaceInfo` in the owning `MooseMesh`.

**Comment by rwcarlsen** [comment]
I'm all for using composition and functional style functioning functions as much as possible - which is a fancy way of saying "yes - I also would prefer making the fvfaceinterface functions just global/helper functions rather than having a separate interface everyone inherits from".  An FV namespace sounds great.  It's always much much easier to remove or rename a namespace than to introduce one later.

File: framework/include/mesh/MooseMesh.h
Comment (lindsayad) [reply framework/include/mesh/MooseMesh.h:1010]: Done (hopefully)

File: framework/include/mesh/MooseMesh.h
Comment (lindsayad) [reply framework/include/mesh/MooseMesh.h]: Done (hopefully)

File: framework/include/problems/DisplacedProblem.h
Comment (lindsayad) [reply framework/include/problems/DisplacedProblem.h:272]: So what I've done is eliminate the `cached_jacobian_contributions*` members in `Assembly` and I've deprecated methods with "Contribution" in the name because it is very confusing to me.

File: framework/include/variables/MooseVariableField.h
Comment (lindsayad) [reply framework/include/variables/MooseVariableField.h:91]: Ok I was able to eliminate this API by adding a private `setDofValueHelper` to `AuxKernel`.

**Comment by lindsayad** [comment]
Ugh, I think this is going to need distributed mesh fixes included in #15338 and #15655

**Comment by rwcarlsen** [comment]
ouch.  - so we have a triple stack - of libmesh, global dofs, this.  And it will be a quadruple stack when I start pushing up commits for fv array vars.

**Comment by rwcarlsen** [comment]
When you get a chance - let's get this rebased and conflicts resolved.

**Comment by lindsayad** [comment]
Hammering away on tests in #15338. Once those are passing, I can ensure that tests on this PR are working

**Comment by rwcarlsen** [comment]
You've done some more work on this since I last looked.  It looks like you've got a few unchecked boxes in the description still.  Where does this stand - is it ready for review or are you still tinkering with stuff?

**Comment by lindsayad** [comment]
Probably should migrate at least the first TO-DO to an issue. That shouldn't hold this up.

I started working on generic block restriction on top of this branch, and am close to done. We can postpone that to another follow-on PR though.

**Comment by rwcarlsen** [comment]
I'll do a final/full review on this sometime today.

**Comment by lindsayad** [comment]
> I started working on generic block restriction on top of this branch, and am close to done. We can postpone that to another follow-on PR though.

Do you want me to postpone this? I'm about ready to push it. Finishing my last test

**Comment by rwcarlsen** [comment]
I'll do my review as soon as you get that pushed up.

**Comment by lindsayad** [comment]
WTF these tests all pass on my machine and they are valgrind clean... and they just straightup seg fault on CIVET?!

**Comment by rwcarlsen** [comment]
Beautiful code.

**Comment by lindsayad** [comment]
Ah this is a configuration difference...

**Comment by ajacquey** [comment]
Hi Alex,

I just tested your branch and I can confirm that your changes fix #15329 
Amazing work, I can't wait to test the new capabilities.

Thanks!

**Comment by lindsayad** [comment]
> Amazing work, I can't wait to test the new capabilities.

Awesome. Thank you for taking the time to check

File: framework/include/variables/MooseVariableFV.h
Comment (rwcarlsen) [inline framework/include/variables/MooseVariableFV.h:298]: This doc comment should explain _what_ is uncorrected (i.e. cross-diffusion correction).

File: framework/include/variables/MooseVariableFV.h
Comment (rwcarlsen) [inline framework/include/variables/MooseVariableFV.h:292]: This doc comment should explain that you are getting a cross-diffusion-corrected face gradient.

File: framework/include/variables/MooseVariableFV.h
Comment (rwcarlsen) [inline framework/include/variables/MooseVariableFV.h:561]: what coefficients?

**Review by rwcarlsen — CHANGES_REQUESTED** [review]
Okay this is all I could muster today.  You wouldn't think it from the (low) number of comments, but I read through this code for almost 2 hours.

File: framework/include/loops/ComputeFVFluxThread.h
Comment (lindsayad) [reply framework/include/loops/ComputeFVFluxThread.h:479]: I didn't like doing this either. But how do I handle this?
````
[FVFluxKernels]
  [fluid_kernel]
    type = FluidKernel
    rho = rho
    mu = mu
    block = 0
  []
[]

[Materials]
  [fluid_mat]
    type = GenericConstantMaterial
    block = 0
    prop_names = 'rho mu'
    prop_values = '1 1'
  []
  [solid_mat]
    type = GenericConstantMaterial
    prop_names = 'rho'
    prop_values = '1000'
  []
[]
```

File: framework/src/fvbcs/FVFluxBC.C
Comment (lindsayad) [reply framework/src/fvbcs/FVFluxBC.C:90]: It is a correctness issue. If the variable isn't defined on a subdomain, then `Moose::adOffset` doesn't really have any meaning. Moreover, if you don't check here, you would have to check the size of `_local_ke` before doing this:
```
    _local_ke(0, 0) = residual.derivatives()[ad_offset];
```
This was the reason for my segfaults originally on CIVET when I didn't have any on my local machines. You don't get here with global indexing. If you do, and you don't have the subdomain check then your indexing into column 0 is out of bounds

File: framework/include/loops/ComputeFVFluxThread.h
Comment (lindsayad) [reply framework/include/loops/ComputeFVFluxThread.h:479]: Moreover, even if I did, how do I tell `fluid_mat` to only compute `mu` on the interface but not `rho`?

File: framework/src/problems/FEProblemBase.C
Comment (lindsayad) [reply framework/src/problems/FEProblemBase.C:6903]: It was not explicitly broken before. It's possible that it worked; probably did.

But yea with the ghosting of materials, e.g. possibly reiniting a face material from the neighbor subdomain in the same place that I'm reiniting face materials from the elem subdomain, I don't know if I understand how to swap stateful properties in that case. Maybe it's fine. It makes my head hurt when I think about this ghosting stuff

File: framework/src/variables/MooseVariableFV.C
Comment (lindsayad) [reply framework/src/variables/MooseVariableFV.C:53]: Because of all this stuff:
```
template <typename T>
void
Assembly::copyShapes(MooseVariableField<T> & v)
{
  phi(v).shallowCopy(v.phi());
  gradPhi(v).shallowCopy(v.gradPhi());
  if (v.computingSecond())
    secondPhi(v).shallowCopy(v.secondPhi());
}

void
Assembly::copyShapes(unsigned int var)
{
  auto & v = _sys.getVariable(_tid, var);
  if (v.fieldType() == Moose::VarFieldType::VAR_FIELD_STANDARD)
  {
    auto & v = _sys.getActualFieldVariable<Real>(_tid, var);
    copyShapes(v);
  }
  else if (v.fieldType() == Moose::VarFieldType::VAR_FIELD_ARRAY)
  {
    auto & v = _sys.getActualFieldVariable<RealEigenVector>(_tid, var);
    copyShapes(v);
  }
  else if (v.fieldType() == Moose::VarFieldType::VAR_FIELD_VECTOR)
  {
    auto & v = _sys.getActualFieldVariable<RealVectorValue>(_tid, var);
    copyShapes(v);
    if (v.computingCurl())
      curlPhi(v).shallowCopy(v.curlPhi());
  }
  else
    mooseError("Unsupported variable field type!");
}
```
Since I'm adding support for coupling FV variables into FE, then I need all these methods to be implemented and not error at run-time when called

File: framework/include/variables/MooseVariableFV.h
Comment (lindsayad) [reply framework/include/variables/MooseVariableFV.h:439]: :+1:

File: framework/include/loops/ComputeFVFluxThread.h
Comment (rwcarlsen) [reply framework/include/loops/ComputeFVFluxThread.h:479]: Why wouldn't you want to use the neighboring subdomain's material there?  If you didn't want to use the neighboring subdomain's material in this case, then rho represents something different in each subdomain and the user should correspondingly name it differently - like rho_fluid and rho_solid.  But when they are the same name, I think we should treat them as one continuous material prop field.

Comment (rwcarlsen) [reply framework/include/loops/ComputeFVFluxThread.h:479]: Wouldn't you still need mu on the interface? - you just need to ghost it instead of using the property from the neighboring subdomain (like you can with rho).

Comment (lindsayad) [reply framework/include/loops/ComputeFVFluxThread.h:479]: >  What if I have 4 subdomains, and for one material property I want a different constant generic property value on each?

and

>  then rho represents something different in each subdomain and the user should correspondingly name it differently - like rho_fluid and rho_solid

seem inconsistent.

> Why wouldn't you want to use the neighboring subdomain's material there?

Because the `fluid_kernel` is restricted to block 0. The user clearly intends to only use material properties from block 0. However if they are using Dirichlet BCs, they are going to need to ghost value for rho on the neighboring side of the interface between blocks 0 and 1. So I want to make sure that `fluid_mat` gets evaluated on the neighbor side of the interface between block 0 and 1. However, if I am also evaluating a material property `rho` in `solid_mat`, then there is a race condition there.

Comment (lindsayad) [reply framework/include/loops/ComputeFVFluxThread.h:479]: > Wouldn't you still need mu on the interface? - you just need to ghost it instead of using the property from the neighboring subdomain (like you can with rho).

So you're talking about whole new code for material property evaluation that would happen outside of the `Material` class. This is probably what you were thinking when we talked on the phone. That's reasonable. *But* I still think we should make it impossible for the `fluid_kernel` to use the `solid_mat` value for `rho` because that's clearly not what the user intended (because of his/her choice for block restriction). What I've implemented here accomplishes that.

Comment (rwcarlsen) [reply framework/include/loops/ComputeFVFluxThread.h:479]: Yes but what you implemented here also makes impossible a very common use case.  I'm saying expecting users to supply a different property name for regions with different physics doesn't seem unreasonable.  However, the user might not have written the kernels.  And might unwittingly use a neighbor material property when they expected ghosting.  Hmmm.  Downsides to all approaches here.  I'll need to think on it a bit.

Comment (lindsayad) [reply framework/include/loops/ComputeFVFluxThread.h:479]: > However, the user might not have written the kernels.

The issue is with the suppliers right? So there definitely could be a problem if the user wants to use existing `Materials` that hard-code a property name. My solution for that would be to go update those `Materials` and make them more flexible. 

> Yes but what you implemented here also makes impossible a very common use case

I do agree. I think it's totally reasonable for a user to define 'k' on multiple blocks. It's at the core of the physical concept of a material. I'm defining a thermal conductivity. Thermal conductivity is denoted by 'k'. I should be able to define 'k' for every material in my simulation.

But in the end I'd rather make the user do more work in the input file than make it possible for the simulation to do something wrong/unexpected.

If there's a straightforward implementation that accomplishes both: 1) guaranteed safety and 2) user convenience, then I'm all for it. I haven't thought of it, but maybe you will.

Ping me on slack if you want to talk about it over the phone.

**Comment by lindsayad** [comment]
Nice job iterating back and forth @rwcarlsen . With https://github.com/idaholab/moose/pull/15825/commits/7a1971dd06d4b14aeb6d22496db6e3bb6cc784c3 I'm quite happy that we can define only one variable and one kernel and do different diffusivities on different  subdomains

Screenshot from 2020-10-21 15-53-39

File: framework/include/variables/MooseVariableFV.h
Comment (lindsayad) [reply framework/include/variables/MooseVariableFV.h:439]: I actually use `getElemValue` and `getNeighborValue` to build the Rhie-Chow interpolation in NSFV. I can make the others private though

File: framework/include/variables/MooseVariableFV.h
Comment (lindsayad) [reply framework/include/variables/MooseVariableFV.h:447]: I actually think is bad design. These coefficients are really specific to the Rhie-Chow interpolation and I put them here only because I didn't quickly think of another way to store them in a single cache of data. I'm going to strip this out for now and if I really don't think of a better way to do this I can add this back in the NSFV PR (but I really should think of a better way)

**Review by rwcarlsen — DISMISSED** [review]
This looks good.  The relaxation on the material block restriction error is a good step in the right direction.  Over time we can hopefully get this relaxed even more (#15991).

**Comment by lindsayad** [comment]
#15338 is merged, and now this is rebased. Will need a re-approval

**Comment by cpgr** [comment]
Just wanted to say that I have been using this branch for some work this week and everything is working well with different materials on different subdomains, so thanks a lot for all your effort!

**Comment by lindsayad** [comment]
That's really nice to hear. Thank you very much for your feedback Chris!

**Comment by cpgr** [comment]
One thing I just noticed is that I can't do value-based adaptivity. For example, if you use a `ValueRangeMarker` and only FV variables, you get an error like `_variable is null in marker. Are you using a finite volume variable?`. Spatial-based adaptivity like `BoxMarker` does work ok though. Hopefully this can be added soon (I can try but I have no idea where to start for this).

**Comment by lindsayad** [comment]
Haha, well at least the error message is pretty spot-on!

**Comment by cpgr** [comment]
4jz4cq

**Comment by lindsayad** [comment]
@cpgr try your value-based markers now that I've added https://github.com/idaholab/moose/pull/15825/commits/9893a195bcfb73e720280a762e70a1a70903a3a4

**Comment by cpgr** [comment]
Yep, I think it is working now. Thanks again!

**Comment by rwcarlsen** [comment]
Looks like you've got a porous flow adaptivity test crashing.

**Comment by rwcarlsen** [comment]
It looks like a few griffin tests are crashing and magpie fails to compile.

**Comment by lindsayad** [comment]
> It looks like a few griffin tests are crashing and magpie fails to compile.

The Griffin tests have been crashing ad infinitum. I will look into a fix for Magpie

**Comment by cpgr** [comment]
I'll take a look at that porous flow test to see if I can work out why it crashed

**Comment by lindsayad** [comment]
Ok everything passing except Griffin

**Review by rwcarlsen — APPROVED** [review]

**Pull request merged**
